units/data_cleanup_testresults.go

```diff
@@ -0,0 +1,140 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/testresult"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/mongodb/grip/sometimes"
+)
+
+const testResultsCleanupJobName = "data-cleanup-testresults"
+
+func init() {
+	registry.AddJobType(testResultsCleanupJobName, func() amboy.Job {
+		return makeTestResultsCleanupJob()
+	})
+}
+
+type dataCleanupTestResults struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	env evergreen.Environment
+}
+
+func makeTestResultsCleanupJob() *dataCleanupTestResults {
+	j := &dataCleanupTestResults{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    testResultsCleanupJobName,
+				Version: 0,
+			},
+		},
+	}
+
+	j.SetDependency(dependency.NewAlways())
+
+	return j
+}
+
+func NewTestResultsCleanupJob(ts time.Time) amboy.Job {
+	j := makeTestResultsCleanupJob()
+	j.SetID(fmt.Sprintf("%s.%s", testResultsCleanupJobName, ts.Format(TSFormat)))
+	return j
+}
+
+func (j *dataCleanupTestResults) Run(ctx context.Context) {
+	defer j.MarkComplete()
+	startAt := time.Now()
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	flags, err := evergreen.GetServiceFlags()
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+
+	if flags.BackgroundCleanup {
+		grip.InfoWhen(sometimes.Percent(evergreen.DegradedLoggingPercent), message.Fields{
+			"job_type": testResultsCleanupJobName,
+			"job_id":   j.ID(),
+			"message":  "disaster recovery backups disabled, also disabling cleanup",
+		})
+		return
+	}
+
+	var (
+		offsets   []int
+		batches   int
+		numDocs   int
+		timeSpent time.Duration
+	)
+
+	totalDocs, _ := j.env.DB().Collection(testresult.Collection).EstimatedDocumentCount(ctx)
+
+	lock := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer func() { j.AddError(recovery.HandlePanicWithError(recover(), nil, "double batch deletes")) }()
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					break
+				default:
+					if time.Since(startAt) >= 45*time.Second {
+						break
+					}
+					opStart := time.Now()
+					offset := rand.Intn(7)
+					num, err := testresult.DeleteWithLimit(ctx, j.env, time.Now().Add(time.Duration(-(offset+365)*24)*time.Hour), 100*1000)
```

Comment:
I would expect the deletes to start at the beginning of time, so I don't see what the reason for the offset is.

Reply:
I imagine that they find the point that they start at in the index and then range from there? randomizing one of the bounds reduces the chance that parallel deleteOnes (from different requests start at the same point.

Reply:
If findOne is a clue, then they start at the first (oldest document), so regardless of the offset they'd start in the same place.

Reply:
ok, I can remove the fuzzing, if they clobber, we'll see it in the variability of total deletions.